resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala

```diff
@@ -225,6 +225,18 @@ private[spark] object Config extends Logging {
         "Ensure that major Python version is either Python2 or Python3")
       .createWithDefault("2")
 
+  val KUBERNETES_DRIVER_CONTAINER_NAME =
+    ConfigBuilder("spark.kubernetes.driver.containerName")
```

Comment:
Is there any way to do this via the `containers` container array from the pod template?

Reply:
I prefer to make it explicit, rather than e.g. "pick the first container in the list".

Reply:
If you're saying that manipulating `containers` doesn't give specific control over which name is assigned to which container, then I agree

Reply:
@mccheah it raises interesting question: a user might use this feature to add containers, e.g. side-car proxies, service mesh, etc. In fact, I'd want users to be able to do this. If there *are* multiple containers defined, we may need a convention for identifying which one is driver/executor. Alternatives might include additive-only, like labels, or disallowing, but supporting this seems very desirable.

Reply:
This feature should support using multiple containers, in which case the user needs to specify which container is running the Spark process. Using a configuration option for that seems like the most straightforward solution.

Reply:
I'm having mixed feelings about utility of convention vs configuration on this. Mostly just because the purpose of the PR is to prevent further spark configurations. If the feature leans on convention that first container in the list is to be the spark driver/executor, how confusing is that likely to be?

Reply:
I'm leaning towards agreeing with @erikerlandson. If we setup the convention that the first container in case of multiple containers in the pod template is the one running the Spark driver/executor, it makes more sense to me to just specify and use the name of the container in the pod template.

Reply:
I generally prefer being more explicit about selecting the container name. Selecting the first container has the danger of silent failures. Especially since we currently hardcode the container name right now, somebody might write their pod templates with a driver container name `spark-kubernetes-driver` but not list it as the first, and then be confused why their spark applications are failing? But I can see how this could be confusing because it assumes a strong relationship between the spark conf and the template.

Reply:
I'm sensing some confusion over the purpose of `spark.kubernetes.driver.containerName` - is it to select the container out of the `containers` array in the template, or is it to *set* that name? Since this PR is creating `containerName` spark conf, I'm assuming it is no longer hard-coded, and we're exposing it to user control.

IIUC, the current concept is that the user *must* declare a container name, *and* also must specify it via `spark.kubernetes.driver.containerName`; If not, I'm not sure what the default behavior is.

I agree that leaning on convention (spark container is the first one) can cause problems, although the same problems would happen if they specify the wrong container name in the config. Clear documentation would be important regardless.  In general, user supplied templates are a "power-user" feature, and some increased opportunities for error are inherent here.

Reply:
Similarly, what if the user gives the driver container name in the driver template but forgets to specify `spark.kubernetes.driver.containerName`? Requiring users to set the container name explicitly and additionally through another config key sounds a bit awkward to me, particularly when there's only the Spark container itself.

Reply:
Is there confusion because there's an _existing_ configuration option also? I think the existing configuration option sets the driver container name when no yml is specified. But perhaps the interpretation of this configuration value should change when the pod template is provided.

Reply:
Yes, the semantic of this key will change with the template option. So with the template, there're two sources of driver container name, and we need to resolve conflicts in certain cases, e.g., if the name is specified in only one source or if the two sources mismatch.

Reply:
Wait, I don't think there's an existing configuration option. I believe the container name is currently just hard coded in `Constants`. 

However, I see your concern about using `spark.kubernetes.driver.containerName` for multiple purposes. In that case, it's definitely easier to reason about with fewer moving pieces, and it sounds like simplest is best. I'll remove the container name configuration and just stick with first container = driver container. This just means that the only way to configure the container name will be through a pod template. I'll also clarify in the docs

Reply:
You are right. We were confusing the one for driver pod name with this one.